feat(builder): allow environments for windows and panes

zappolowski opened the pull request:
This fixes #832.

Open question: what's the intended/desired fallback behavior?

1. (Currently implemented) If a pane doesn't have a `environment` configuration it uses the `environment` configuration of the window. This means that in order to change one environment variable (compared to the window) the full `environment` configuration needs to be duplicated. This also makes it possible to unset some variable if needed (compared to 2.) by skipping it int the configuration.
2. The `environment` of a pane is `ChainMap`ped with the one of the window (and possibly the session?) which means that the final `environment` contains all variables from both pane and window with the pane value taking precedence over the one from the window. This would avoid the possible duplication of 1. but makes it harder to unset an variable (in case an empty value doesn't suffice).
3. Something else? Proposals welcome.

Tests are currently failing as the check for old tmux versions isn't implemented yet.

tony commented:
@zappolowski Hola! I will be able to give this a look over the weekend

tony commented:
Rebased 0b129c6

tony commented:
Rebased against d63f9f5

tony commented:
Things that would help users see the value of the contribution:

- An [example](https://tmuxp.git-pull.com/configuration/examples.html)
- Mention the new functionality the changelog
- Note the functionality of pane + window environments are only in versions 1.19 and beyond

I notice it's in a draft: When it's in a state where it's ready for review (that may be now) let me know

zappolowski commented:
It's still in draft as there's still the question what the intended behavior is (see PR message). I've added another commit (could be fixuped if needed later) to demonstrate this.

I've skipped doing the chore works until the implementation is settled.

tony commented:
@zappolowski 

> It's still in draft as there's still the question what the intended behavior is (see PR message). I've added another commit (could be fixuped if needed later) to demonstrate this.

> 1. (Currently implemented) If a pane doesn't have a `environment` configuration it uses the `environment` configuration of the window. This means that in order to change one environment variable (compared to the window) the full `environment` configuration needs to be duplicated. This also makes it possible to unset some variable if needed (compared to 2.) by skipping it int the configuration.

For this PR, this is best. It's explicit, and it allows us to add support for inheriting / merging the environmental options at a later time.

> 2. The `environment` of a pane is `ChainMap`ped with the one of the window (and possibly the session?) which means that the final `environment` contains all variables from both pane and window with the pane value taking precedence over the one from the window. This would avoid the possible duplication of 1. but makes it harder to unset an variable (in case an empty value doesn't suffice).
> 3. Something else? Proposals welcome.

In a version 2, I think we can introduce a system for this. Can we create an GH issue / discussion to discuss this further?

P.S. This is optiona, but do you use slack? I have a channel for libtmux / tmuxp development. I can invite you

zappolowski commented:
@tony I've added the feature to `CHANGES` and updated the examples (also removed the last commit as this will be done later).

> P.S. This is optional, but do you use slack? I have a channel for libtmux / tmuxp development. I can invite you

Sure.

tony commented:
Invited you to slack via the email in your git commits. If that doesn't work email me (via the email in the git log)

tony commented:
@zappolowski Forecast: I may not be able to get to this until next weekend (I'd ideally like to give it my undivided attention)

I am juggling a few time sensitive tasks outside tmuxp at once

tony commented:
At [3a3362d](https://github.com/tmux-python/tmuxp/pull/845/commits/3a3362d381121a7788cab325969218b74f63ea03)

tmux 3.2

zsh

Locally:

```
====================================================== FAILURES ======================================================
_____________________________________________ test_environment_variables _____________________________________________
tests/workspace/test_builder.py:364: in test_environment_variables
    assert pane.capture_pane()[1] == "PANE"
E   AssertionError: assert 't% echo $FOO' == 'PANE'
E     - PANE
E     + t% echo $FOO
        builder    = <tmuxp.workspace.builder.WorkspaceBuilder object at 0x7f169f4f3390>
        no_overrides_win = Window(@2 1:no_overrides, Session($1 libtmux_t4q8jr92))
        pane       = Pane(%4 Window(@4 3:pane_overrides, Session($1 libtmux_t4q8jr92)))
        pane_overrides_win = Window(@4 3:pane_overrides, Session($1 libtmux_t4q8jr92))
        session    = Session($1 libtmux_t4q8jr92)
        window_overrides_win = Window(@3 2:window_overrides, Session($1 libtmux_t4q8jr92))
        workspace  = {'environment': {'FOO': 'SESSION', 'PATH': '/tmp'}, 'session_name': 'test env vars', 'start_directory': '/tmp/pytest-o...es': [{'shell_command': []}, {'environment': {'FOO': 'PANE'}, 'shell_command': []}], 'window_name': 'both_overrides'}]}
============================================== short test summary info ===============================================
FAILED tests/workspace/test_builder.py::test_environment_variables - AssertionError: assert 't% echo $FOO' == 'PANE'
===================================== 1 failed, 199 passed, 4 skipped in 33.80s ======================================
```

tony commented on tests/workspace/test_builder.py:
```diff
@@ -340,9 +345,68 @@ def test_environment_variables(session):
+    assert session.getenv("PATH") == "/tmp"
+
+    no_overrides_win = session.windows[0]
+    pane = no_overrides_win.panes[0]
+    pane.send_keys("echo $FOO")
```
in re: https://github.com/tmux-python/tmuxp/pull/845#issuecomment-1320908658

I needed this locally for the tests to pass

```suggestion
    pane = no_overrides_win.panes[0]
    time.sleep(0.3)  # To let shells with slow startup catch up
    pane.send_keys("echo $FOO")
```

zappolowski replied:
That positions seems a bit arbitrary for the sleep. How about waiting after the full session is built (line 346)?

tony replied:
@zappolowski Can you push a change so I can see what you mean?

tony replied:
> That positions seems a bit arbitrary for the sleep. How about waiting after the full session is built (line 346)?

I think I get what you mean (though why not just commit it?) [I can't commit it](https://github.com/tmux-python/tmuxp/pull/845#issuecomment-1321236255).

tony replied:
Confirmed on my side this works:

```diff
diff --git a/tests/workspace/test_builder.py b/tests/workspace/test_builder.py
index 2852c2f2..98463f9a 100644
--- a/tests/workspace/test_builder.py
+++ b/tests/workspace/test_builder.py
@@ -344,6 +344,7 @@ def test_environment_variables(session):

     builder = WorkspaceBuilder(sconf=workspace)
     builder.build(session)
+    time.sleep(0.3)

     assert session.getenv("FOO") == "SESSION"
     assert session.getenv("PATH") == "/tmp"
```

tony replied:
Well done!

tony commented:
@zappolowski Are you [allowing commits to the PR?](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)

zappolowski commented:
> @zappolowski Are you [allowing commits to the PR?](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)

Usually not (at least intentionally) as my commits are signed which won't work when other people are committing as well. So, I have to rebase anyhow to get the signatures correct.

tony commented:
@zappolowski Is this ready to go?

tony commented:
> Usually not (at least intentionally) as my commits are signed which won't work when other people are committing as well. So, I have to rebase anyhow to get the signatures correct.

Noted!

tony commented on src/tmuxp/workspace/builder.py:
```diff
@@ -346,12 +347,22 @@ def iter_create_windows(self, session, append=False):
+            if has_gte_version("3.0"):
+                environment = panes[0].get("environment", wconf.get("environment"))
+            else:
+                logging.warning(
```
Wouldn't this show the warning for every window, even when the user hasn't set an environment?

tony replied:
Maybe fine to remove the warning? (providing we move it to libtmux?)

re: https://github.com/tmux-python/tmuxp/pull/845/files#r1030592815

zappolowski replied:
You're right ... somehow one check went missing. Re-added them (and left the warning in place as explained in the other comment).

tony commented on src/tmuxp/workspace/builder.py:
```diff
@@ -418,11 +429,21 @@ def get_pane_shell():
+                        "Cannot set environment for new pane. "
+                        "You need tmux 3.0 or newer for this."
+                    )
+                    environment = {}
```
I suppose we can remove the warning (and move it to libtmux perhaps?)

zappolowski replied:
Libtmux already has warnings in that case (see [here](https://github.com/tmux-python/libtmux/blob/6a03ad929368544cf07762799cdf4b1c821619ce/src/libtmux/session.py#L273) and [here](https://github.com/tmux-python/libtmux/blob/6a03ad929368544cf07762799cdf4b1c821619ce/src/libtmux/window.py#L532)). My idea is to (maybe in the future) provide a better warning (e.g. file name and line number etc.). Also, I wouldn't rely on a lib to provide warnings in such cases (besides that they probably also use different loggers, which might give different outputs).

zappolowski commented:
I've reworked the warnings issued for older versions of tmux to be a bit more precise and helpful and adjusted the tests accordingly.

tony reviewed (approved):
Good!

Minor note: I may make some changes around the logging.  It's a bit difficult to explain in the PR due to the nature of the change.

tony commented:
@zappolowski Now live on tmuxp v1.19.0a0 ([PyPI](https://pypi.org/project/tmuxp/1.19.0a0/), [GitHub Release](https://github.com/tmux-python/tmuxp/releases/tag/v1.19.0a0), Changes on [Docs](https://github.com/tmux-python/tmuxp/blob/v1.19.0a0/CHANGES#tmuxp-119x-unreleased), [Site](http://localhost:8031/history.html#tmuxp-1-19-x-unreleased))

Let's dogfood this!

tony commented:
@zappolowski This is now live in [1.19.0](https://pypi.org/project/tmuxp/1.19.0/)!
